### PR TITLE
Clean up log_events.go

### PR DESCRIPTION
**Pull request by coventry**
I keep finding this code hard to follow, whenever I come back to it. The goal of
this PR is to document the role of log_events.go's logic and data, and to
re-organize it a bit so its knowledge is a bit more centralized.

### CODE REVIEW

File: core/store/models/common.go [hunk 1]
```diff
@@ -183,32 +183,53 @@ func (j JSON) Bytes() []byte {
...
 	default:
-		return JSON{}, errors.New("can only add to JSON objects or null")
+		return nil, errors.New("can only add to JSON objects or null")
 	}
+	return output, nil
+}
 
-	bytes, err := json.Marshal(output)
+// mapToJSON returns m as a JSON object, or errors
+func mapToJSON(m map[string]interface{}) (JSON, error) {
```
Comment (felder-cl) [inline core/store/models/common.go]: NewJSONFromMap(..) ;)

File: core/store/models/common.go [hunk 2]
```diff
@@ -183,32 +183,53 @@ func (j JSON) Bytes() []byte {
...
+// Add returns a new instance of JSON with the new value added.
+func (j JSON) Add(insertKey string, insertValue interface{}) (JSON, error) {
+	return j.MultiAdd([]KV{{insertKey, insertValue}})
+}
+
+// KV represents a key/value pair to be added to a JSON object
+type KV struct {
+	Key   string
+	Value interface{}
+}
```
Comment (felder-cl) [inline core/store/models/common.go]: Why not use a map instead?

File: core/store/models/log_events.go [hunk 3]
```diff
@@ -85,55 +86,75 @@ var topicFactoryMap = map[common.Hash]logRequestParser{
...
-// possibly be entered on a RunLog or a ServiceAgreementExecutionLog. There is the ID,
-// hex encoded and the ID zero padded.
-func TopicFiltersForRunLog(logTopics []common.Hash, jobID *ID) [][]common.Hash {
-	return [][]common.Hash{logTopics, {IDToTopic(jobID), IDToHexTopic(jobID)}}
+// LogBasedChainlinkJobInitiators are initiators which kick off a log-specified
+// chainlink job when an appropriate ethereum log is received.
+// (InitiatorFluxMonitor kicks off a job when a log is emitted, but the JobID is
+// not taken from the log)
+var LogBasedChainlinkJobInitiators = []string{InitiatorRunLog, InitiatorEthLog,
+	InitiatorServiceAgreementExecutionLog}
```
Comment (felder-cl) [inline core/store/models/log_events.go]: Maybe splitting these into single lines would make it easier to inspect?

File: core/store/models/log_events.go [hunk 4]
```diff
@@ -85,55 +86,75 @@ var topicFactoryMap = map[common.Hash]logRequestParser{
...
+
+// topicsForInitiatorsWhichRequireJobSpecTopic are the log topics which kick off
```
Comment (felder-cl) [inline core/store/models/log_events.go]: topics should be capitalized

File: core/store/models/log_events.go [hunk 5]
```diff
@@ -85,55 +86,75 @@ var topicFactoryMap = map[common.Hash]logRequestParser{
...
+
+// initiationRequiresJobSpecId is true if jobs initiated by the given
+// initiatiatorType require that their initiating logs match their JobSpecIDs.
+func initiationRequiresJobSpecID(initiatorType string) bool {
+	_, ok := TopicsForInitiatorsWhichRequireJobSpecIDTopic[initiatorType]
+	return ok
+}
+
+// jobSpecIDTopics lists the ways jsID could be represented as a log topic. This
+// allows log subscriptions to respond to all possible representations.
```
Comment (felder-cl) [inline core/store/models/log_events.go]: maybe? // jobSpecIDTopics lists the ways its ID could be represented as a log topic. This

File: core/store/models/log_events.go [hunk 6]
```diff
@@ -85,55 +86,75 @@ var topicFactoryMap = map[common.Hash]logRequestParser{
...
+	}
 }
 
 // FilterQueryFactory returns the ethereum FilterQuery for this initiator.
-func FilterQueryFactory(i Initiator, from *big.Int) (ethereum.FilterQuery, error) {
-	q := ethereum.FilterQuery{
-		FromBlock: from,
-		Addresses: utils.WithoutZeroAddresses([]common.Address{i.Address}),
-	}
+func FilterQueryFactory(i Initiator, from *big.Int) (q ethereum.FilterQuery, err error) {
```
Comment (felder-cl) [inline core/store/models/log_events.go]: I'm of the impression defining the return values this way is only kosher for short functions.  I can't seem to find a link on me.

File: core/store/models/log_events.go [hunk 7]
```diff
@@ -154,21 +175,21 @@ type LogRequest interface {
...
 
 // InitiatorLogEvent encapsulates all information as a result of a received log from an
-// InitiatorSubscription.
+// InitiatorSubscription, and acts as a base struct for other log-initiated events
 type InitiatorLogEvent struct {
 	Log       eth.Log
 	Initiator Initiator
 }
 
+var _ LogRequest = InitiatorLogEvent{} // InitiatorLogEvent implements LogRequest
```
Comment (felder-cl) [inline core/store/models/log_events.go]: :+1:

File: core/store/models/log_events.go [hunk 8]
```diff
@@ -398,21 +410,15 @@ func (p parseRunLog0original) parseJSON(log eth.Log) (JSON, error) {
...
-	if err != nil {
-		return js, err
-	}
-
-	return js.Add("functionSelector", OracleFullfillmentFunctionID0original)
+	return js.MultiAdd([]KV{
+		{"address", log.Address.String()},
+		{"dataPrefix", bytesToHex(data[:idSize])},
+		{"functionSelector", OracleFullfillmentFunctionID0original},
+	})
```
Comment (felder-cl) [inline core/store/models/log_events.go]: Looks like a map ;)

File: core/store/models/log_events.go [hunk 9]
```diff
@@ -85,55 +86,75 @@ var topicFactoryMap = map[common.Hash]logRequestParser{
...
-		FromBlock: from,
-		Addresses: utils.WithoutZeroAddresses([]common.Address{i.Address}),
-	}
+func FilterQueryFactory(i Initiator, from *big.Int) (q ethereum.FilterQuery, err error) {
+	q.FromBlock = from
+	q.Addresses = utils.WithoutZeroAddresses([]common.Address{i.Address})
 
-	switch i.Type {
-	case InitiatorEthLog:
+	switch {
```
Comment (j16r) [inline core/store/models/log_events.go]: This code would be clearer if the safe path returned a new literal FilterQuery and the case blocks just set a `FromBlock` variable. IMHO.

File: core/store/models/common.go (hunk 1 above)
Comment (coventry) [reply core/store/models/common.go]: Thanks. :)

File: core/store/models/common.go (hunk 2 above)
Comment (coventry) [reply core/store/models/common.go]: You're right, that's much cleaner. I've re-arranged it that way.

File: core/store/models/log_events.go (hunk 4 above)
Comment (coventry) [reply core/store/models/log_events.go]: Thanks.

File: core/store/models/log_events.go (hunk 6 above)
Comment (coventry) [reply core/store/models/log_events.go]: What's the drawback to this?

File: core/store/models/log_events.go (hunk 5 above)
Comment (coventry) [reply core/store/models/log_events.go]: Changed it to `...the ways jobSpecID could be...`, and changed the variable name accordingly.

**Comment by coventry** [comment]
> don't see much of logic centralization

More knowledge centralization, and that's mostly `LogBasedChainlinkJobInitiators` and `TopicsForInitiatorsWhichRequireJobSpecIDTopic`, so that we don't have those lists explicitly elsewhere in the code.

File: core/store/models/log_events.go (hunk 9 above)
Comment (coventry) [reply core/store/models/log_events.go]: You're right, that's much nicer. Thanks.

File: core/store/models/log_events.go (hunk 6 above)
Comment (felder-cl) [reply core/store/models/log_events.go]: I suggest we just stick to whats recommended unless there's good reason.  It saves time and makes things consistent.

File: core/store/models/common.go [hunk 10]
```diff
@@ -183,32 +183,51 @@ func (j JSON) Bytes() []byte {
...
-	bytes, err := json.Marshal(output)
+// NewJSONFromMap returns m as a JSON object, or errors
+func NewJSONFromMap(m map[string]interface{}) (JSON, error) {
+	bytes, err := json.Marshal(m)
 	if err != nil {
 		return JSON{}, err
 	}
-
 	return JSON{Result: gjson.ParseBytes(bytes)}, nil
 }
```
Comment (felder-cl) [inline core/store/models/common.go]: I have code prepared to introduce NewJSON(interface{}), I think this generalized version will take car of this case.

File: core/store/models/common.go [hunk 11]
```diff
@@ -183,32 +183,51 @@ func (j JSON) Bytes() []byte {
...
+func (j JSON) MultiAdd(keyValues jsonObject) (JSON, error) {
+	output, err := j.AsMap()
+	if err != nil {
+		return JSON{}, err
+	}
+	for key, value := range keyValues {
+		output[key] = value
+	}
+	return NewJSONFromMap(output)
+}
```
Comment (felder-cl) [inline core/store/models/common.go]: What was trying to be done by having a public method task a private value?

Comment (felder-cl) [reply core/store/models/common.go]: Is this function heavily used in later PRs?  I don't see much use case of it in this PR, and it could be done with multiple Add()'s.

Comment (felder-cl) [reply core/store/models/common.go]: Also, another option would be to add a JSON.Merge(JSON) function which would be more generalized.
